substrate/frame/identity/src/lib.rs

```diff
@@ -886,6 +966,180 @@ pub mod pallet {
+			UsernameAuthorities::<T>::insert(
+				&authority,
+				AuthorityProperties { suffix, allocation },
+			);
+			Self::deposit_event(Event::AuthorityAdded { authority });
+			Ok(())
+		}
+
+		/// Remove `authority` from the username authorities.
+		#[pallet::call_index(16)]
+		#[pallet::weight(T::WeightInfo::remove_username_authority())]
+		pub fn remove_username_authority(
+			origin: OriginFor<T>,
+			authority: AccountIdLookupOf<T>,
+		) -> DispatchResult {
+			T::UsernameAuthorityOrigin::ensure_origin(origin)?;
+			let authority = T::Lookup::lookup(authority)?;
+			ensure!(
+				UsernameAuthorities::<T>::contains_key(&authority),
+				Error::<T>::NotUsernameAuthority
+			);
+			UsernameAuthorities::<T>::remove(&authority);
+			Self::deposit_event(Event::AuthorityRemoved { authority });
+			Ok(())
+		}
+
+		/// Set the username for `who`. Must be called by a username authority.
+		#[pallet::call_index(17)]
+		#[pallet::weight(T::WeightInfo::set_username_for())]
+		pub fn set_username_for(
+			origin: OriginFor<T>,
+			who: AccountIdLookupOf<T>,
+			username: Vec<u8>,
+			signature: Option<T::OffchainSignature>,
+		) -> DispatchResult {
+			// Ensure origin is a Username Authority and has an allocation. Decrement their
+			// allocation by one.
+			let sender = ensure_signed(origin)?;
+			let suffix = UsernameAuthorities::<T>::try_mutate(
+				&sender,
+				|maybe_authority| -> Result<Suffix, DispatchError> {
+					let properties =
+						maybe_authority.as_mut().ok_or(Error::<T>::NotUsernameAuthority)?;
+					ensure!(properties.allocation > 0, Error::<T>::NoAllocation);
+					properties.allocation.saturating_dec();
+					Ok(properties.suffix.clone())
+				},
+			)?;
+
+			// Ensure that the username only contains allowed characters. We already know the suffix
+			// does.
+			let username_length = username.len().saturating_add(suffix.len()) as u32;
+			Self::validate_username(&username, Some(username_length))?;
+
+			// Concatenate the username with suffix and cast into a BoundedVec. Should be infallible
+			// since we already ensured it is below the max length.
+			let mut full_username =
+				Vec::with_capacity(username.len().saturating_add(suffix.len()).saturating_add(1));
+			full_username.extend(username);
+			full_username.extend(b".");
+			full_username.extend(suffix);
+			let bounded_username =
+				Username::try_from(full_username).map_err(|_| Error::<T>::InvalidUsername)?;
+
+			// Usernames must be unique. Ensure it's not taken.
+			ensure!(
+				!AccountOfUsername::<T>::contains_key(bounded_username.clone()),
+				Error::<T>::UsernameTaken
+			);
+
+			// Insert or queue.
+			let who = T::Lookup::lookup(who)?;
+			if let Some(s) = signature {
+				// Account has pre-signed an authorization. Verify the signature provided and grant
+				// the username directly.
+				let encoded = Encode::encode(&bounded_username.to_vec());
+				Self::validate_signature(&encoded, &s, &who)?;
```

Comment:
I guess there is no real threat of the authority re-using these signatures? Otherwise the account nonce could be included to improve it a bit.

Reply:
How would they re-use them? It would still be for the same username, and it'd be bad UX if an account can't transact between signing a message and the authority submitting it.